ledger/complete/ledger.go

```diff
@@ -229,10 +266,44 @@ func (l *Ledger) Set(update *ledger.Update) (newState ledger.State, trieUpdate *
 
 	state := update.State()
 	l.logger.Info().Hex("from", state[:]).
-		Hex("to", newRootHash[:]).
+		Hex("to", newState[:]).
 		Int("update_size", update.Size()).
 		Msg("ledger updated")
-	return ledger.State(newRootHash), trieUpdate, nil
+	return newState, trieUpdate, nil
+}
+
+func (l *Ledger) set(trieUpdate *ledger.TrieUpdate) (newState ledger.State, err error) {
+
+	resultCh := make(chan error)
+
+	if l.trieUpdateCh == nil {
+		go func() {
+			_, _, err := l.wal.RecordUpdate(trieUpdate)
+			resultCh <- err
+		}()
```

Comment:
how does concurrent write to a disk WAL file work? Is it ultimately only one file writer writing at a time, and all other writers are blocked waiting?

If so, we might want to make WAL writing sequential, in order to ensure all trie updates of a block are written to a segment file before writing the next.

Reply:
The code you highlighted is existing code currently running in mainnet.  With this PR, that block of code is only used in tests for `Ledger`, etc without having to run a MockCompactor.  Getting rid of it is on my mind but it can be done later because it doesn't impact the main goal of this PR.

> how does concurrent write to a disk WAL file work? Is it ultimately only one file writer writing at a time, and all other writers are blocked waiting?

Yes.  `wal` package uses write lock so only one writing at a time.

> If so, we might want to make WAL writing sequential, in order to ensure all trie updates of a block are written to a segment file before writing the next.

WAL writing is already sequential by sending `trieUpdate` over a channel so `Compactor` processes `trieUpdate` in sequence.  https://github.com/onflow/flow-go/blob/65f20620dc38ec35efce75e407a6ee241410c091/ledger/complete/ledger.go#L284-L289